zkevm-circuits/src/evm_circuit/execution/call.rs

```diff
@@ -0,0 +1,718 @@
+        let is_success = cb.query_bool();
+
+        // Use rw_counter of the step which triggers next call as its call_id.
+        let callee_call_id = cb.curr.state.rw_counter.clone();
+
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+        let mut reversion_info = cb.reversion_info(None);
+        let [caller_address, is_static, depth] = [
+            CallContextFieldTag::CallerAddress,
+            CallContextFieldTag::IsStatic,
+            CallContextFieldTag::Depth,
+        ]
+        .map(|field_tag| cb.call_context(None, field_tag));
+
+        cb.range_lookup(depth.expr(), 1024);
+
+        // Lookup values from stack
+        cb.stack_pop(gas_word.expr());
+        cb.stack_pop(callee_address_word.expr());
+        cb.stack_pop(value.expr());
+        cb.stack_pop(cd_offset.expr());
+        cb.stack_pop(cd_length.expr());
+        cb.stack_pop(rd_offset.expr());
+        cb.stack_pop(rd_length.expr());
+        cb.stack_push(is_success.expr());
+
+        // Recomposition of random linear combination to integer
+        let callee_address =
+            from_bytes::expr(&callee_address_word.cells[..N_BYTES_ACCOUNT_ADDRESS]);
+        let gas = from_bytes::expr(&gas_word.cells[..N_BYTES_GAS]);
+        let gas_is_u64 = IsZeroGadget::construct(cb, sum::expr(&gas_word.cells[N_BYTES_GAS..]));
+        let cd_address = MemoryAddressGadget::construct(cb, cd_offset, cd_length);
+        let rd_address = MemoryAddressGadget::construct(cb, rd_offset, rd_length);
+        let memory_expansion = MemoryExpansionGadget::construct(
+            cb,
+            cb.curr.state.memory_word_size.expr(),
+            [cd_address.address(), rd_address.address()],
+        );
+
+        // Add callee to access list
+        let is_warm_access = cb.query_bool();
+        let is_warm_access_prev = cb.query_bool();
+        cb.account_access_list_write(
+            tx_id.expr(),
+            callee_address.clone(),
+            is_warm_access.expr(),
+            is_warm_access_prev.expr(),
+            Some(&mut reversion_info),
+        );
+
+        // Propagate rw_counter_end_of_reversion and is_persistent
+        let mut callee_reversion_info = cb.reversion_info(Some(callee_call_id.expr()));
+        cb.require_equal(
+            "callee_is_persistent == is_persistent ⋅ is_success",
+            callee_reversion_info.is_persistent(),
+            reversion_info.is_persistent() * is_success.expr(),
+        );
+        cb.condition(is_success.expr() * (1.expr() - reversion_info.is_persistent()), |cb| {
+            cb.require_equal(
+                "callee_rw_counter_end_of_reversion == rw_counter_end_of_reversion - (state_write_counter + 1)",
+                callee_reversion_info.rw_counter_end_of_reversion(),
+                reversion_info.rw_counter_of_reversion(),
+            );
+        });
+
+        // Verify transfer
+        let value_is_zero = IsZeroGadget::construct(cb, sum::expr(&value.cells));
```

Comment:
Can't you add `gas_is_u64` and `value` and check that the result is 0? On that way with one gadget instantiation you cover both gadget results at once.

Just an idea.

Reply:
But call needs the bools `has_value` and `gas_is_u64` respectively for different constraint, if check them together we can't tell their individual value.

Reply:
Hmmm I see! The idea was that if we constrain the addition of the values to be zero. We know that both values are zero. But that's in fact not true as we can overflow with modulo arith and go back to 0. 
Feel free to ingore this.